app/src/app/classes/stix/stix-object.ts

```diff
@@ -565,6 +564,103 @@ export abstract class StixObject extends Serializable {
+    if (!this.firstInitialized || !this.supportsNamespace) return;
+    let sub = this.getNamespaceID(restAPIService).subscribe({
+      next: (val) => {
+        this.attackID = val
+      },
+      complete: () => sub.unsubscribe()
+    });
+  }
+
+  public getNamespaceID(restAPIConnector): Observable<any> {
+    let prefix = ''; // i.e. 'PRE-T'
+    let count = ''; // i.e. '1234'
+    let copyID = this.attackID;
+    this.attackID = '(generating ID)';
+    return restAPIConnector.getOrganizationNamespace().pipe(
+      map(namespaceSettings => namespaceSettings),
+      switchMap(organizationNamespace => {
+        if (organizationNamespace && organizationNamespace.hasOwnProperty('prefix')) {
+          if (organizationNamespace['prefix']) prefix += (organizationNamespace['prefix'] + '-');
+          count = organizationNamespace['range_start'];
+          count = (Number(count) > 0 ? count : 1).toString().padStart(4, '0'); // make sure ID has 4 digits i.e. 0 -> 0001 (note: padStart() is unsupported in IE)
+        }
+        let accessor = this.attackType == "group" ? restAPIConnector.getAllGroups() :
+                       this.attackType == "mitigation" ? restAPIConnector.getAllMitigations() :
+                       this.attackType == "software" ? restAPIConnector.getAllSoftware() :
+                       this.attackType == "tactic" ? restAPIConnector.getAllTactics() :
+                       this.attackType == "technique" ? restAPIConnector.getAllTechniques() :
+                       this.attackType == "data-source" ? restAPIConnector.getAllDataSources() :
+                       this.attackType == "matrix" ? restAPIConnector.getAllMatrices() : null;
+        // Find all other objects that have this prefix and range, and set ID to the most recent and unique ID possible
+        if (accessor) {
+          prefix += this.attackIDValidator.format.includes('#') ? this.attackIDValidator.format.split('#')[0] : '';
+          return accessor.pipe(map((objects) => {
+            let substr = prefix + count.substr(0,3); // i.e. look for 'PRE-T123x' matches
+            let filtered = objects['data'].filter((obj) => obj.attackID.startsWith(substr));
+
+            let reg = new RegExp("\\d{4}");
+            filtered = filtered.sort((a, b,) => {
+              a = a.attackID.match(reg);
+              b = b.attackID.match(reg);
+              return a[0] - b[0] // check which 4-digit ID is greater
+            });
+
+            let latest = filtered.length > 0 ? filtered.pop().attackID.match(reg)[0] : '0';
+            latest = Number(latest)
+            count = (Number(count) > latest ? count : latest + 1).toString().padStart(4, '0');
+
+            // Generate next available ID (assumes user had already clicked generate once and that's why the attackID already exists)
+            if (copyID && copyID.replace(/[A-Z]+-/i,'').endsWith(count)) {
+              count = (Number(count) + 1).toString().padStart(4, '0');
+            }
```

Comment:
I'm a little confused by this part of the code, is this adding +1 again if the generated ID already existed?

Reply:
Yep, this was to +1 if the user just clicked the button twice. But I guess in the edge case this was trying to solve (where another user might be saving another object with the same ID) this is not necessary since this generate function fetches all the objects of this type to check for conflicting IDs anyway. I'll remove this.